### ISSUE TITLE
[Feature] Implement the DeepSeek-V4-Flash DSpark drafter

### ISSUE THREAD

**Issue by high-cloud** [hw-native-sys/pypto-lib#935]
### Summary

Implement the DeepSeek-V4-Flash DSpark drafter in
`models/deepseek_v4_flash_dspark/`.

The drafter consists of:

1. A three-layer parallel draft backbone using checkpoint layers
   `mtp.0`, `mtp.1`, and `mtp.2`.
2. A rank-256 sequential Markov head that emits seven speculative tokens
   from one anchor token.

The first implementation is scoped to PyPTO-Lib with synthetic golden
validation. Serving integration and checkpoint conversion are tracked
separately.

### Motivation / Use Case

The existing `models/deepseek_v4_flash_dspark/` directory contains the target
model operators needed to verify `1 + 7` tokens, but it does not contain the
DSpark drafter that produces those seven speculative tokens.

Without the drafter, the directory cannot exercise or measure the complete
DSpark speculative-decoding path described by the DeepSeek-V4-Flash checkpoint
and the vLLM/vLLM-Ascend integration.

This issue provides an independently reviewable implementation and validation
track for P6 of #905.

#### Public programs

Implement two independently compilable JIT/L3 programs:

1. `draft_backbone`
   - Project the three selected target hidden states into the drafter hidden
     state.
   - Build an anchor-first query block:
     `[anchor_token, noise_token × 6]`.
   - Execute three DSpark draft layers.
   - Maintain an independent SWA KV cache for each draft layer.
   - Produce `head_hidden[B, 7, hidden_size]`.

2. `markov_sample`
   - Apply the final RMSNorm and shared target LM head.
   - Compute the rank-256 Markov bias.
   - Greedily sample seven tokens sequentially.
   - Feed each sampled token into the next Markov transition.
   - Produce `draft_token_ids[B, 7]`.

Keeping the backbone and sampler separate limits the public tensor ABI and lets
the two stages be validated independently.

#### Model contract

| Item | Value |
|---|---|
| Draft layers | `mtp.0`, `mtp.1`, `mtp.2` |
| Target hidden layers | 40, 41, 42 |
| Drafter query width | 7 |
| Noise token ID | 128799 |
| Markov rank | 256 |
| Attention type | Sliding-window attention |
| Query-block attention | Non-causal within the seven-token block |
| Public output | `draft_token_ids[B, 7]` |

For every draft layer, each query position must attend to:

- the valid historical SWA window;
- the current anchor context KV;
- all seven query-block KVs.

The seven-token query block is therefore non-causal internally. Draft-layer KV
caches must remain isolated from one another.

#### Layer composition

Each of the three draft layers executes:

```text
HC pre
  -> SWA attention
  -> HC post
  -> MoE
```

MoE synchronization state must be cleared between layers, and each layer must
use a distinct communication epoch.

After the third layer:

```text
HC head
  -> RMSNorm
  -> target LM head
  -> sequential Markov bias and greedy sampling
```

### Implementation Checklist

- [ ] Add DSpark drafter constants and shape assertions to `config.py`.
- [ ] Implement the target-hidden projection and anchor/noise query layout.
- [ ] Implement non-causal SWA metadata for the seven-token query block.
- [ ] Compose the three draft layers with stacked layer weights and KV caches.
- [ ] Implement the final HC head and RMSNorm path.
- [ ] Implement rank-256 Markov projection and sequential greedy sampling.
- [ ] Add synthetic PyTorch goldens and focused component tests.
- [ ] Compile both public programs on the supported A2/A3 target.
- [ ] Validate synthetic golden results on a real A2/A3 NPU.
- [ ] Record supported batch sizes and remaining performance work in this issue.

### Validation

Synthetic-golden coverage must include:

- target-hidden projection;
- anchor/noise query construction;
- non-causal visibility across all seven query positions;
- per-layer KV-cache isolation;
- short context and full SWA window;
- page-boundary and ring-wrap cases;
- three-layer intermediate hidden states;
- final head hidden states and base logits;
- Markov bias and final token IDs;
- a causality check proving that changing an earlier sampled token can change
  later Markov outputs without changing the base logits.

Exercise dynamic single-card batches `{4, 8, 12, 16}`. At minimum, run `B=4`
and `B=16` through the real-NPU synthetic-golden harness.

### Alternatives Considered

- Reusing the existing MTP drafter unchanged: rejected because DSpark generates
  a non-causal token block through three parallel draft layers and then performs
  sequential Markov refinement. This is not the MTP-1 execution contract.
- Implementing the whole drafter as one public JIT program: deferred because the
  combined model state and cache arguments risk exceeding the runtime tensor
  argument limit and make component-level validation harder.
- Adding serving integration in the same change: deferred so the PyPTO kernel
  contract can be reviewed and validated first.

### Additional Context

Initial scope does not include:

- serving scheduler or proposer integration;
- target-model verification or rejection sampling;
- bonus-token commit behavior;
- the confidence head;
- official checkpoint download or conversion;
- performance acceptance criteria.

References:

- DSpark paper: https://arxiv.org/abs/2607.05147
- DeepSeek-V4-Flash-DSpark checkpoint:
  https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-DSpark
- Umbrella deployment issue: #905

**Comment by high-cloud** [comment]
Implementation progress (local branch `codex/deepseek-v4-dspark-drafter`):

Completed:
- added the DSpark constants and host-side contracts;
- implemented target-hidden preparation and the padded anchor/noise query block;
- implemented layer-specific, non-causal SWA metadata and DSpark attention;
- composed the three-layer backbone with isolated KV caches, MoE signal clears, and epochs 1/2/3;
- implemented the final HC-head output and the sequential rank-256 Markov sampler;
- added focused contract tests and updated the model index.

Validation:
- `pytest -q tests/contract/test_dspark_drafter_contract.py`: 12 passed;
- ruff, headers, English-only, and `git diff --check`: passed;
- prepare, metadata, and attention component compile checks: passed;
- Markov compile at checkpoint vocabulary size 129280: passed;
- real A2/A3 NPU synthetic golden:
  - B=4 on device 2: PASS (`task_20260810_214936_396176027237`);
  - B=16 on device 3: PASS (`task_20260810_215123_411758930262`).

Remaining before closing this issue:
- add/exercise a full multi-rank synthetic harness for the composed three-layer backbone;
- performance characterization and serving integration remain outside this first kernel-only slice, as scoped in the issue.
